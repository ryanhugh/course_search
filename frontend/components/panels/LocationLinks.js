/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';
import PropTypes from 'prop-types';

import macros from '../macros';


// Calculate the Google Maps links from a given section.
// This is used in both the mobile section panel and the desktop section panel.

function LocationLinks(props) {
  const elements = props.locations.map((location, index, locations) => {
    let buildingName;
    if (location.match(/\d+\s*$/i)) {
      buildingName = location.replace(/\d+\s*$/i, '');
    } else {
      buildingName = location;
    }

    let optionalComma = null;
    if (index !== locations.length - 1) {
      optionalComma = ', ';
    }

    if (location.toUpperCase() === 'TBA' || location.toUpperCase() === 'LOCATION TBA') {
      if (locations.length > 1) {
        return null;
      }

      return location;
    }

    if (location.toUpperCase() === 'BOSTON DEPT') {
      return (
        <span key='Boston DEPT'>
          TBA &nbsp;(Boston Campus)
        </span>
      );
    }

    // The <a> tag needs to be on one line, or else react will insert spaces in the generated HTML.
    // And we only want spaces between these span elements, and not after the location and the comma.
    // eg YMCA, Hurting Hall and not YMCA , Hurting Hall
    return (
      <span key={ location }>
        <a target='_blank' rel='noopener noreferrer' href={ `https://maps.google.com/?q=${macros.collegeName} ${buildingName}` }>
          {location}
        </a>
        {optionalComma}
      </span>
    );
  });

  return (
    <span>
      {elements}
    </span>
  );
}


LocationLinks.propTypes = {
  locations: PropTypes.array.isRequired,
};


export default LocationLinks;
