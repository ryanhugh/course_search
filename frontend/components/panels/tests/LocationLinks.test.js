/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import mockData from './mockData';
import LocationLinks from '../LocationLinks';

Enzyme.configure({ adapter: new Adapter() });


it('should render a section', () => {
  const result = shallow(<LocationLinks locations={ mockData.cs1210.sections[0].getLocations() } />);
  expect(result).toMatchSnapshot();
});


it('should render another section', () => {
  const result = shallow(<LocationLinks locations={ mockData.cs0210.sections[0].getLocations() } />);
  expect(result).toMatchSnapshot();
});
