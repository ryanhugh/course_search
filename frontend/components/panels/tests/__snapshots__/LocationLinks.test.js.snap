// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a section 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocationLinks
    locations={
      Array [
        "Snell Engineering Center 108",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span>
          <a
            href="https://maps.google.com/?q=Northeastern University Snell Engineering Center "
            rel="noopener noreferrer"
            target="_blank"
          >
            Snell Engineering Center 108
          </a>
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "Snell Engineering Center 108",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <a
              href="https://maps.google.com/?q=Northeastern University Snell Engineering Center "
              rel="noopener noreferrer"
              target="_blank"
            >
              Snell Engineering Center 108
            </a>,
            null,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Snell Engineering Center 108",
              "href": "https://maps.google.com/?q=Northeastern University Snell Engineering Center ",
              "rel": "noopener noreferrer",
              "target": "_blank",
            },
            "ref": null,
            "rendered": "Snell Engineering Center 108",
            "type": "a",
          },
          null,
        ],
        "type": "span",
      },
    ],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span>
            <a
              href="https://maps.google.com/?q=Northeastern University Snell Engineering Center "
              rel="noopener noreferrer"
              target="_blank"
            >
              Snell Engineering Center 108
            </a>
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "Snell Engineering Center 108",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <a
                href="https://maps.google.com/?q=Northeastern University Snell Engineering Center "
                rel="noopener noreferrer"
                target="_blank"
              >
                Snell Engineering Center 108
              </a>,
              null,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Snell Engineering Center 108",
                "href": "https://maps.google.com/?q=Northeastern University Snell Engineering Center ",
                "rel": "noopener noreferrer",
                "target": "_blank",
              },
              "ref": null,
              "rendered": "Snell Engineering Center 108",
              "type": "a",
            },
            null,
          ],
          "type": "span",
        },
      ],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`should render another section 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocationLinks
    locations={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
    },
    "ref": null,
    "rendered": Array [],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
      },
      "ref": null,
      "rendered": Array [],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
